x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -352,10 +359,20 @@ public void onFailure(Exception e) {
     }
 
     private Tuple<String, Settings> loadMappingAndSettingsSourceFromTemplate() {
-        final byte[] template = TemplateUtils.loadTemplate("/" + SECURITY_TEMPLATE_NAME + ".json",
-                Version.CURRENT.toString(), SecurityIndexManager.TEMPLATE_VERSION_PATTERN).getBytes(StandardCharsets.UTF_8);
+        final byte[] template = TemplateUtils.loadTemplate("/" + SECURITY_TEMPLATE_NAME + ".json", Version.CURRENT.toString(),
+                SecurityIndexManager.TEMPLATE_VERSION_PATTERN).getBytes(StandardCharsets.UTF_8);
         PutIndexTemplateRequest request = new PutIndexTemplateRequest(SECURITY_TEMPLATE_NAME).source(template, XContentType.JSON);
-        return new Tuple<>(request.mappings().get("doc"), request.settings());
+        final byte[] typeMappingSource = request.mappings().get(MapperService.SINGLE_MAPPING_NAME).getBytes(StandardCharsets.UTF_8);
+        Map<String, Object> typeMappingSourceMap = XContentHelper
+                .convertToMap(new BytesArray(typeMappingSource, 0, typeMappingSource.length), true, XContentType.JSON).v2();
+        XContentBuilder builder = null;
+        try {
+            builder = XContentFactory.jsonBuilder();
+            builder.value(typeMappingSourceMap.get(MapperService.SINGLE_MAPPING_NAME));
+            return new Tuple<>(Strings.toString(builder), request.settings());
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToRuntime(e);
+        }
```

Comment:
The above code looks to me as we are rewriting the type of the mappings in the template to `_doc`. However the template already has `_doc` as a type name in the template so this shouldn't be necessary? Am I missing something?

Reply:
The problem here is a combination of 2 things:
1. Indices that were upgraded from 6.x will have `"doc"` type, but we want indices that were created on 7.x to have `_doc` as the type.
2. We use the `PutIndexTemplateRequest` in order to parse our "template" source and convert it into something we can use for `PutMappingRequest`, and that doesn't quite do what we need here.

`PutIndexTemplateRequest.getMappings()` always returns _typed_ mappings. There's special code in the rest action (via `RestCreateIndexAction.prepareMappings`) that accepts untyped input and converts it to a typed request using `_doc`.

So, what we get out from 
```
request.mappings().get(MapperService.SINGLE_MAPPING_NAME)
```
will be typed (that is, it has a top-level "_doc" key) so if we feed that verbatim into the `PutMappingRequest` it will be treated as a type'd request, rather than an untyped one.

The code above rips out that top level `"_doc"` key so that `PutMappingRequest` uses the untyped path, and updates the single type, which might be `_doc` (for an index created in 7.x) or `doc` (for upgraded indices).

There's probably a neater way to rip out that top level `_doc` type from the mapping - this XContent code seems more complex than I'd expect - but I think it's necessary to do something like this.

The other options would be:
1. Don't convert to "_doc" at all, and keep using "doc" on new, 7.x indices.
2. Replace the top level key with whatever the actual type is, but that MetaDataMappingService already has code to do that for us, so that seems redundant.
3. Convert the `MetaDataMappingService` to accept typed mappings, on SINGLE_MAPPING_TYPE even if that type is not the real underlying type name (this isn't terrible, but it seems likely to cause problems elsewhere).

Reply:
Thanks for explaining, I get it now. I think the current approach is the right one.

Reply:
I can't find a way to make this parsing/conversion substantially better. I would suggest calling `convertToMap`  using the mapping `String` without converting it to bytes, but that's a minor change.

I think it needs to be commented and/or split into smaller methods so it's clear what's going on. The logic seems fine, but it takes some effort to follow why we're doing it.